File: convert_maze.py
from PIL import Image, ImageDraw

def convert_maze(image_path):
    # Load the image
    image = Image.open(image_path)
    draw = ImageDraw.Draw(image)
    
    # Get image dimensions
    img_width, img_height = image.size
    
    # Calculate grid square dimensions
    column_count = 29
    row_count = column_count
    grid_width = img_width // column_count
    grid_height = img_height // row_count
    
    # Initialize JavaScript list
    js_list = []

    for row in range(row_count):
        row_list = []
        for col in range(column_count):
            # Calculate the top-left corner of the grid square
            x1 = col * grid_width
            y1 = row * grid_height
            
            # Calculate the bottom-right corner of the grid square
            x2 = x1 + grid_width - 1
            y2 = y1 + grid_height - 1
            
            # Make sure we are within the image boundaries
            x2 = min(x2, img_width - 1)
            y2 = min(y2, img_height - 1)
            
            # Calculate the center of the grid square
            center_x = (x1 + x2) // 2
            center_y = (y1 + y2) // 2
            
            # Get the color of the center pixel
            center_pixel = image.getpixel((center_x, center_y))
            
            # Check if the center pixel is white
            is_white = ((center_pixel[0] + center_pixel[1] + center_pixel[2])/3) >= 200

            # Add 1 for white and 0 for non-white to the row list
            row_list.append(1 if is_white else 0)
            
            # Draw the number at the center of the grid square
            text_color = "black" if is_white else "white"

            if is_white:
                draw.text((center_x-1, center_y-5), str(1), fill=text_color)
            else:
                draw.text((center_x-1, center_y-5), str(0), fill=text_color)

        # Add the row list to the JavaScript list
        js_list.append(row_list)

    js_list.insert(0, "var mazeData = [")
    js_list.append("];")

    # Save the annotated image
    labeled_maze = "mazes/labeled_maze.png"
    image.save(labeled_maze)

    return js_list, labeled_maze, column_count

if "__main__" == __name__:
    js_list, maze_image_path, column_count = convert_maze("mazes/maze4.png")

    print("-" * 33, " <-- Maze data --> ", "-" * 33, "\n")

    for row in js_list:
        if len(row) >= column_count:
            print(f"{row},")
        else:
            print(row)

    print("-" * 33, " <-- Maze data --> ", "-" * 33, "\n")

    

File: convertMaze.js
console.log("convertMaze.js loaded");
// Load the image into a canvas
async function loadImage(src) {
    console.log("Entering loadImage with src:", src);
    return new Promise((resolve) => {
        const img = new Image();
        img.onload = () => {
            console.log("Image loaded successfully");
            resolve(img);
        };
        img.onerror = () => {
            console.log("Failed to load image from:", src);
        };
        img.src = src;
    });
}

async function convertMaze(imagePath, colCount) {
    console.log("Called: convertMaze(", imagePath, colCount, ")");
    window.sharedData = null;
    const canvas = document.createElement('canvas');
    canvas.willReadFrequently = true;
    const ctx = canvas.getContext('2d', { willReadFrequently: true });
    // ctx.willReadFrequently(true);
    const img = await loadImage(imagePath);
    if(!img) {
        console.log("Image loading failed. Exiting convertMaze.");
        return;
    }
    canvas.width = img.width;
    canvas.height = img.height;
    ctx.drawImage(img, 0, 0);

    const imgWidth = img.width;
    const imgHeight = img.height;
    const columnCount = colCount;
    const rowCount = columnCount;
    const gridWidth = Math.floor(imgWidth / columnCount);
    const gridHeight = Math.floor(imgHeight / rowCount);

    console.log("Image dimensions:", imgWidth, "x", imgHeight);
    console.log("Grid dimensions:", gridWidth, "px x", gridHeight, "px");

    let jsList = [];

    for (let row = 0; row < rowCount; row++) {
        let rowList = [];
        for (let col = 0; col < columnCount; col++) {
            const x1 = col * gridWidth;
            const y1 = row * gridHeight;
            let x2 = x1 + gridWidth - 1;
            let y2 = y1 + gridHeight - 1;

            x2 = Math.min(x2, imgWidth - 1);
            y2 = Math.min(y2, imgHeight - 1);

            const centerX = Math.floor((x1 + x2) / 2);
            const centerY = Math.floor((y1 + y2) / 2);

            const pixelData = ctx.getImageData(centerX, centerY, 1, 1).data;
            const red = pixelData[0];
            const green = pixelData[1];
            const blue = pixelData[2];
            
            const isWhite = (red + green + blue) / 3 >= 230;
            rowList.push(isWhite ? 1 : 0);
        }
        jsList.push(rowList);
    }

    // jsList.unshift("[");
    // jsList.push("]");
    // console.log("Generated jsList:", jsList);
    // 
    window.sharedData = jsList;
    return { jsList };
}

// Uncommented the example usage for further testing if needed
// convertMaze('image-path', 10).then((result) => {
//     console.log("Result from convertMaze:", result);
// });
